### PR TITLE
Raise an error if we try to have interfaces with auto properties on constructor-less types

### PR DESCRIPTION
**Pull request by edgarfgp**
## Description

- Adds a unit test for verify https://github.com/dotnet/fsharp/issues/13690
- Raise a new error if we try to have interfaces with auto properties on constructor-less types

## Checklist

- [x] Test cases added
- [x] Release notes entry updated

### CODE REVIEW

**Comment by edgarfgp** [comment]
This is ready

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 1]
```diff
@@ -4386,6 +4393,14 @@ module TcDeclarations =
                 members |> List.tryPick (function 
                     | SynMemberDefn.ImplicitCtor (ctorArgs = SynSimplePats.SimplePats _ as spats) -> Some spats
                     | _ -> None)
+                                
+            // Raise a new error if we try to have an interface implementation with auto properties on constructor-less types,
+            if not implementedAutoProperties.IsEmpty then
+                for mem in implementedAutoProperties do
+                    match mem with
+                    | SynMemberDefn.AutoProperty(range = m) when implicitCtorSynPats.IsNone ->
```
Comment (T-Gro) [inline src/Compiler/Checking/CheckDeclarations.fs:4401]:
```suggestion
                    | SynMemberDefn.AutoProperty( isStatic = false; range = m) when implicitCtorSynPats.IsNone ->
```

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 2]
```diff
@@ -4386,6 +4393,14 @@ module TcDeclarations =
                 members |> List.tryPick (function 
                     | SynMemberDefn.ImplicitCtor (ctorArgs = SynSimplePats.SimplePats _ as spats) -> Some spats
                     | _ -> None)
+                                
+            // Raise a new error if we try to have an interface implementation with auto properties on constructor-less types,
+            if not implementedAutoProperties.IsEmpty then
+                for mem in implementedAutoProperties do
```
Comment (T-Gro) [inline src/Compiler/Checking/CheckDeclarations.fs:4399]:
```suggestion
                for mem in implementedAutoProperties do
```

File: tests/FSharp.Compiler.ComponentTests/Diagnostics/Records.fs [hunk 3]
```diff
@@ -91,4 +91,23 @@ let t3 (x: RecTy) (a: AnotherNestedRecTy) = { x with D.C.c = { a with A = 3; B =
...
+let ``Error when implementing interface with auto property in record type``() =
+    FSharp """
+type Foo =
+  abstract member X : string with get, set
+  abstract GetValue: unit -> string
+
+type FooImpl = 
+  { name: string }
+  interface Foo with
+    member val X = "" with get, set
```
[the remark this answers is not in the document]
Comment (T-Gro) [reply tests/FSharp.Compiler.ComponentTests/Diagnostics/Records.fs:106]: Like this? https://github.com/fsharp/fslang-suggestions/issues/68

[the remark this answers is not in the document]
Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/Diagnostics/Records.fs:106]: Not really it was more about 

- Allowing records to implement interfaces that content AutoProperties

```fsharp
type Foo =
   abstract member X : string with get, set

type FooImpl = 
   { name: string }
   interface Foo with
     member val X = "Default" with get, set

let foo = { name = "foo" } :> Foo
foo.X <- "NewDefault"
```

Currently fails with `This declaration element is not permitted in an augmentation`


- The existing way of implementing X is 
```fsharp
type FooImpl = 
   { name: string }
   interface Foo with
       member this.X = this.name
       member this.X with set value = () 

let foo = { name = "foo" } :> Foo
foo.X <- "NewDefault"
```

Saw a [comment](https://github.com/dotnet/fsharp/issues/13690#issuecomment-1214935073) from @vzarytovskii that got me thinking about this.

[the remark this answers is not in the document]
Comment (edgarfgp) [reply tests/FSharp.Compiler.ComponentTests/Diagnostics/Records.fs:106]: Anyway this is not really relevant for this PR. I will raise a new issue/feature request to see if there is an interest on this. Resolving the comment.

**Comment by psfinaki** [comment]
Hey @edgarfgp good stuff here. In the PR description, you mention benchmarks, do you still plan to add them?

**Comment by edgarfgp** [comment]
> Hey @edgarfgp good stuff here. In the PR description, you mention benchmarks, do you still plan to add them?

Not applicable in this case. The check was there as part of the new PR template :). I guess I should remove it from the description

**Comment by psfinaki** [comment]
Ah okay, nevermind then :)

**Review by psfinaki — APPROVED** [review]

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 4]
```diff
@@ -4394,6 +4401,13 @@ module TcDeclarations =
                 members |> List.tryPick (function 
                     | SynMemberDefn.ImplicitCtor (ctorArgs = SynSimplePats.SimplePats _ as spats) -> Some spats
                     | _ -> None)
+                                
+            // Raise a new error if we try to have an interface implementation with auto properties on constructor-less types,
+            for mem in implementedAutoProperties do
+                match mem with
+                | SynMemberDefn.AutoProperty( isStatic = false; range = m) when implicitCtorSynPats.IsNone ->
+                    errorR(Error(FSComp.SR.tcAutoPropertyRequiresImplicitConstructionSequence(), m))
+                | _ -> ()
```
Comment (vzarytovskii) [inline src/Compiler/Checking/CheckDeclarations.fs:4410]: I wonder if `implicitCtorSynPats.IsNone` check should wrap the loop, so we don't enumerate properties every time?

Comment (T-Gro) [reply src/Compiler/Checking/CheckDeclarations.fs:4410]: Not just the loop, the whole constructions of `implementedAutoProperties` can be avoided (= allocations saved)

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 5]
```diff
@@ -4353,6 +4350,16 @@ module TcDeclarations =
         match trepr with
         | SynTypeDefnRepr.ObjectModel(kind, members, m) ->
             let members = desugarGetSetMembers members
+            let implementedAutoProperties =
+                members
+                |> List.choose (
```
[the remark this answers is not in the document]
Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs]: Yeah on second look. It is not needed at all. Just moved to `CheckMembersForm` so we do not need this

File: src/Compiler/Checking/CheckDeclarations.fs (hunk 4 above)
Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs]: Moved the check to `CheckMembersForm` where it already check for implicit constructor so this we can use the auto property  check in a more easy way

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 6]
```diff
@@ -4202,12 +4202,24 @@ module TcDeclarations =
...
                 if isAbstractSlot slot then
                     match slot with
                     | SynMemberDefn.AbstractSlot (slotSig = synVal; range = m) ->
                         CheckDuplicatesArgNames synVal m
                     | _ -> ()
+
+                if isInterface slot then
+                    match slot with
+                    | SynMemberDefn.Interface (members= Some defs) when not defs.IsEmpty ->
+                        let autoProps = defs |> List.filter(isAutoProperty)
```
Comment (auduchinok) [inline src/Compiler/Checking/CheckDeclarations.fs:4216]: Do I get it right that it constructs a new list on each member, making the analysis quadratic?

Comment (auduchinok) [reply src/Compiler/Checking/CheckDeclarations.fs:4216]: Sorry, looked at a wrong thing.

Comment (edgarfgp) [reply src/Compiler/Checking/CheckDeclarations.fs:4216]: I think a for loop and checking on the `SynMemberDefn.AutoProperty` would be better that filtering

Comment (auduchinok) [reply src/Compiler/Checking/CheckDeclarations.fs]: Happy to help!

**Pull request merged**